pvlib/spectrum.py

```diff
@@ -0,0 +1,366 @@
+                             surface_pressure, precipitable_water, ozone,
+                             optical_thickness, scattering_albedo, dayofyear):
+    """
+    Calculate transmittance factors from Section 2 of Bird and Riordan 1984.
+    """
+    wavelength = _SPECTRL2_COEFFS['wavelength'][:, np.newaxis]
+    vapor_coeff = _SPECTRL2_COEFFS['water_vapor_absorption'][:, np.newaxis]
+    ozone_coeff = _SPECTRL2_COEFFS['ozone_absorption'][:, np.newaxis]
+    mixed_coeff = _SPECTRL2_COEFFS['mixed_absorption'][:, np.newaxis]
+
+    # ET spectral irradiance correction for earth-sun distance seasonality.
+    # Note that we only want the distance correction coefficient, so set
+    # solar_constant=1:
+    earth_sun_distance_correction = \
+        pvlib.irradiance.get_extra_radiation(dayofyear, method='spencer',
+                                             solar_constant=1)  # Eq 2-2, 2-3
+    # Rayleigh scattering
+    # note: 101300 is used for consistentcy with reference; can't use
+    # atmosphere.get_absolute_airmass because it uses 101325
+    airmass = relative_airmass * surface_pressure / 101300
+    wavelength_um = wavelength / 1000
+    rayleigh_transmittance = np.exp(
+        # Note: the report uses 1.335 but spectrl2_2.c uses 1.3366
+        # -airmass / (wavelength_um**4 * (115.6406 - 1.335 / wavelength_um**2))
+        -airmass / (wavelength_um**4 * (115.6406 - 1.3366 / wavelength_um**2))
+    )  # Eq 2-4
+
+    # Aerosol scattering and absorption, Eq 2-6
+    aerosol_transmittance = np.exp(-optical_thickness * relative_airmass)
+
+    # Water vapor absorption, Eq 2-8
+    aWM = vapor_coeff * precipitable_water * relative_airmass
+    vapor_transmittance = np.exp(-0.2385 * aWM / (1 + 20.07 * aWM)**0.45)
+
+    # Ozone absorption
+    ozone_max_height = 22
+    h0_norm = ozone_max_height / 6370
+    ozone_mass_numerator = (1 + h0_norm)
+    ozone_mass_denominator = np.sqrt(cosd(apparent_zenith)**2 + 2 * h0_norm)
+    ozone_mass = ozone_mass_numerator / ozone_mass_denominator  # Eq 2-10
+    ozone_transmittance = np.exp(-ozone_coeff * ozone * ozone_mass)  # Eq 2-9
+
+    # Mixed gas absorption, Eq 2-11
+    aM = mixed_coeff * airmass
+    # Note: the report uses 118.93, but spectrl2_2.c uses 118.3
+    # mixed_transmittance = np.exp(-1.41 * aM / (1 + 118.93 * aM)**0.45)
+    mixed_transmittance = np.exp(-1.41 * aM / (1 + 118.3 * aM)**0.45)
+
+    # split out aerosol components for diffuse irradiance calcs
+    aerosol_scattering = np.exp(
+        -scattering_albedo * optical_thickness * relative_airmass
+    )  # Eq 3-9
+
+    aerosol_absorption = np.exp(
+        -(1 - scattering_albedo) * optical_thickness * relative_airmass
+    )  # Eq 3-10
+
+    return (
+        earth_sun_distance_correction,
+        rayleigh_transmittance,
+        aerosol_transmittance,
+        vapor_transmittance,
+        ozone_transmittance,
+        mixed_transmittance,
+        aerosol_scattering,
+        aerosol_absorption,
+    )
```

Comment:
Although it's a private function, I think would be worth commenting the shapes of the arguments and return values.